src/stream/src/executor/dml_reader.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::pin::Pin;
+use std::task::Poll;
+
+use either::Either;
+use futures::stream::{select_with_strategy, BoxStream, PollNext, SelectWithStrategy};
+use futures::{Stream, StreamExt, TryStreamExt};
+use futures_async_stream::try_stream;
+use risingwave_common::transaction::transaction_message::TxnMsg;
+use risingwave_connector::source::BoxTxnMsgStream;
+
+use crate::executor::error::{StreamExecutorError, StreamExecutorResult};
+use crate::executor::Message;
+
+type ExecutorMessageStream = BoxStream<'static, StreamExecutorResult<Message>>;
+
+type DmlReaderData = StreamExecutorResult<Either<Message, TxnMsg>>;
+type DmlReaderArm = BoxStream<'static, DmlReaderData>;
+type DmlReaderWithPauseInner =
+    SelectWithStrategy<DmlReaderArm, DmlReaderArm, impl FnMut(&mut PollNext) -> PollNext, PollNext>;
+
+/// [`DmlReaderWithPause`] merges two streams, with one receiving barriers (and maybe other types
+/// of messages) and the other receiving dml data only (no barrier). The merged stream can be paused
+/// (`DmlReaderWithPause::pause_stream`) and resumed (`DmlReaderWithPause::resume_stream`).
+/// A paused stream will not receive any data from either original stream until a barrier arrives
+/// and the stream is resumed. The two streams will be polled in a round robin fashion.
+pub(super) struct DmlReaderWithPause {
+    inner: DmlReaderWithPauseInner,
+    /// Whether the source stream is paused.
+    paused: bool,
+}
```

Comment:
Vote for generics here. The implementation here and `StreamReaderWithPause` are very much alike. 👀

Reply:
Will try to use generics here later.